src/compiler/transformers/declarations.ts

```diff
@@ -0,0 +1,1276 @@
+            //   import B = Q;
+            //   import C = B;
+            //   export import D = C;
+            // }
+            // ```
+            // In such a scenario, only Q and D are initially visible, but we don't consider imports as private names - instead we say they if they are referenced they must
+            // be recorded. So while checking D's visibility we mark C as visible, then we must check C which in turn marks B, completing the chain of
+            // dependent imports and allowing a valid declaration file output. Today, this dependent alias marking only happens for internal import aliases.
+            const unconsideredImports: AnyImportSyntax[] = [];
+            while (length(possibleImports)) {
+                const i = possibleImports.shift();
+                if ((isSourceFile(i.parent) ? i.parent : i.parent.parent) !== enclosingDeclaration) { // Filter to only declarations in the current scope
+                    unconsideredImports.push(i);
+                    continue;
+                }
+                // Eagerly transform import equals - if they're not visible, we'll get nothing, if they are, we'll immediately add them since it's complete
+                if (i.kind === SyntaxKind.ImportEqualsDeclaration) {
+                    const result = transformImportEqualsDeclaration(i);
+                    if (result) {
+                        emittedImports.push(result);
+                    }
+                    continue;
+                }
+                // Import declarations, on the other hand, can be partially painted by multiple aliases; so we can see many indeterminate states
+                // until we've marked all possible visibility
+                const result = transformImportDeclaration(i);
+                importDeclarationMap.set("" + getNodeId(i), [i, result]);
+            }
+            // Filtering available imports is the last thing done within a scope, so the possible set becomes those which could not
+            // be considered in the child scope
+            possibleImports = unconsideredImports;
+            // And lastly, we need to get the final form of all those indetermine import declarations from before and add them to the output list
+            // (and remove them from the set to examine for outter declarations)
+            const pairs = arrayFrom(importDeclarationMap.values());
+            for (const [original, replacement] of pairs) {
+                if ((isSourceFile(original.parent) ? original.parent : original.parent.parent) !== enclosingDeclaration) continue; // Filter to only declarations in the current scope
+                importDeclarationMap.delete("" + getNodeId(original));
+                if (replacement) {
+                    emittedImports.push(replacement);
+                }
+            }
+
+            return emittedImports;
+        }
+
+        function visitDeclarationComponents(input: Node): VisitResult<Node> {

```

Comment:
The name here is tripping me up. What is a declaration *component*? Are we just lacking a better name for a fuzzy concept?

Reply:
I mean, the things at the top level are declaration statements. What's a good name for _everything else_?

Reply:
The sub-parts aren't themselves declarations or statements.

Reply:
Isn't subtree just about as overloaded a term as components with respect to an ast transformer? Tbqh, neither term gives you a great idea of what it operates over.